### PR TITLE
Refactor dataload structure and add some engine level method

### PR DESCRIPTION
**Pull request by frankleaf**
### Ⅴ. Special notes for reviews

### CODE REVIEW

File: pkg/controllers/v1alpha1/dataload/implement.go [hunk 1]
```diff
@@ -329,64 +251,21 @@ func (r *DataLoadReconcilerImplement) reconcilePendingDataLoad(ctx reconcileRequ
...
-		r.Recorder.Eventf(&ctx.DataLoad, v1.EventTypeNormal, common.DataLoadJobStarted, "The DataLoad job %s started", jobName)
-		return utils.RequeueAfterInterval(20 * time.Second)
-	}
-
-	// 3. Check running status of the DataLoad job
+	// 2. Check running status of the DataLoad job
+	releaseName := utils.GetDataLoadReleaseName(targetDataload.Name)
+	jobName := utils.GetDataLoadJobName(releaseName)
 	log.V(1).Info("DataLoad chart already existed, check its running status")
 	job, err := utils.GetDataLoadJob(r.Client, jobName, ctx.Namespace)
```
Comment (yangyuliufeng) [inline pkg/controllers/v1alpha1/dataload/implement.go:270]: how about abstracting a function GetDataloadStatus in the runtime engine?
Different runtime perhpas has different implementation methods, we cannot easily get the dataload status by job status

**Review by yangyuliufeng — CHANGES_REQUESTED** [review]
Originally, the logic of dataload was divided into two parts: the general part and the implementation part. The whole reconciledataload was put in the  implementation part

Now, after refactoring dataload structure, due to the addition of the engine level method, most of the logic should be put in the general part, and the implementation part should only include loading the runtime. It's like alluxio_ runtime_ Controller.go and runtime_ controller.go。


原先，dataload的逻辑被分为两部分：通用部分与实现部分，整个ReconcileDataload都放在了具体实现部分
现在，重构dataload逻辑之后，由于增加了engine级的方法，大部分逻辑应该放在通用部分，实现部分的逻辑应该只包括load the runtime。如同alluxio_runtime_controller.go和runtime_controller.go。

File: pkg/ddc/jindo/load_data.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+		if len(defaultImageInfo) < 2 {
+			panic("invalid default dataload image!")
+		} else {
+			imageTag = defaultImageInfo[1]
+		}
+	}
+
+	image := fmt.Sprintf("%s:%s", imageName, imageTag)
+
+	runtime, err := utils.GetJindoRuntime(r.Client, boundedRuntime.Name, boundedRuntime.Namespace)
```
Comment (TrafalgarZZZ) [inline pkg/ddc/jindo/load_data.go:98]: Better to reuse `r.Runtime` with a type conversion here to get the JindoRuntime here

**Review by TrafalgarZZZ — CHANGES_REQUESTED** [review]

File: pkg/ddc/jindo/load_data.go (hunk 2 above)
Comment (frankleaf) [reply pkg/ddc/jindo/load_data.go:98]: thx and I will have a try

**Comment by frankleaf** [comment]
> Originally, the logic of dataload was divided into two parts: the general part and the implementation part. The whole reconciledataload was put in the implementation part
> 
> Now, after refactoring dataload structure, due to the addition of the engine level method, most of the logic should be put in the general part, and the implementation part should only include loading the runtime. It's like alluxio_ runtime_ Controller.go and runtime_ controller.go。
> 
> 原先，dataload的逻辑被分为两部分：通用部分与实现部分，整个ReconcileDataload都放在了具体实现部分
> 现在，重构dataload逻辑之后，由于增加了engine级的方法，大部分逻辑应该放在通用部分，实现部分的逻辑应该只包括load the runtime。如同alluxio_runtime_controller.go和runtime_controller.go。

现在的重构逻辑就是把通用的放在dataload，需要engine自己去实现的都抽出来自己去实现

File: pkg/controllers/v1alpha1/dataload/implement.go (hunk 1 above)
Comment (frankleaf) [reply pkg/controllers/v1alpha1/dataload/implement.go:270]: thx, There are two reasons. First, there are only two kinds of runtimes in a short time, and the judgment behavior of the two runtimes is the same, in order not to cause code redundancy. Second, according to the current data preheating process, this process may be universal. Of course, it makes sense to extract it also

File: pkg/ddc/base/engine.go [hunk 3]
```diff
@@ -41,6 +41,15 @@ type Engine interface {
 
 	// Sync syncs the alluxio runtime
 	Sync(ctx cruntime.ReconcileRequestContext) error
+
+	// LoadData generate dataload values and install helm chart
+	LoadData(ctx cruntime.ReconcileRequestContext, targetDataload datav1alpha1.DataLoad) (err error)
```
Comment (cheyang) [inline pkg/ddc/base/engine.go:46]: How about 

```golang
type DataLoader interface {
        LoadData(ctx cruntime.ReconcileRequestContext, targetDataload datav1alpha1.DataLoad) (err error)

       CheckExistenceOfPath(targetDataload datav1alpha1.DataLoad) (notExist bool, err error)
}

type Engine interface {

       // Sync syncs the alluxio runtime
	Sync(ctx cruntime.ReconcileRequestContext) error
       DataLoader
}
```

File: pkg/controllers/v1alpha1/dataload/dataload_controller.go [hunk 4]
```diff
@@ -101,45 +106,104 @@ func (r *DataLoadReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
...
+		return utils.RequeueAfterInterval(20 * time.Second)
+	}
+	ctx.RuntimeType = boundedRuntime.Type
+
+	switch ctx.RuntimeType {
+	case common.ALLUXIO_RUNTIME:
+		runtime, err := utils.GetAlluxioRuntime(ctx.Client, boundedRuntime.Name, boundedRuntime.Namespace)
+		if err != nil {
+			if utils.IgnoreNotFound(err) == nil {
+				ctx.Log.V(1).Info("The runtime is not found", "runtime", ctx.NamespacedName)
```
Comment (cheyang) [inline pkg/controllers/v1alpha1/dataload/dataload_controller.go:132]: ctx.Log.V(1).Info -> ctx.Log.Info

**Comment by cheyang** [comment]
@frankleaf thanks very much for your PR. It's very important to decouple the definition and implementation. Would you mind to add some unit test here like https://github.com/fluid-cloudnative/fluid/pull/832 . Thank you in advanced.

File: pkg/controllers/v1alpha1/dataload/dataload_controller.go [hunk 5]
```diff
@@ -148,3 +212,25 @@ func (r *DataLoadReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 }
+
+// GetOrCreateEngine gets the Engine
+func (r *DataLoadReconciler) GetOrCreateEngine(
+	ctx cruntime.ReconcileRequestContext) (engine base.Engine, err error) {
+	found := false
+	id := ddc.GenerateEngineID(ctx.NamespacedName)
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	if engine, found = r.engines[id]; !found {
```
Comment (cheyang) [inline pkg/controllers/v1alpha1/dataload/dataload_controller.go:223]: I am wondering how to clean up the engine if the runtime is deleted.

**Review by cheyang — CHANGES_REQUESTED** [review]

**Comment by yangyuliufeng** [comment]
> > Originally, the logic of dataload was divided into two parts: the general part and the implementation part. The whole reconciledataload was put in the implementation part
> > Now, after refactoring dataload structure, due to the addition of the engine level method, most of the logic should be put in the general part, and the implementation part should only include loading the runtime. It's like alluxio_ runtime_ Controller.go and runtime_ controller.go。
> > 原先，dataload的逻辑被分为两部分：通用部分与实现部分，整个ReconcileDataload都放在了具体实现部分
> > 现在，重构dataload逻辑之后，由于增加了engine级的方法，大部分逻辑应该放在通用部分，实现部分的逻辑应该只包括load the runtime。如同alluxio_runtime_controller.go和runtime_controller.go。
> 
> 现在的重构逻辑就是把通用的放在dataload，需要engine自己去实现的都抽出来自己去实现

我指的是dataload_controller.go和implement.go这两个文件的逻辑关系

File: pkg/ddc/base/engine.go (hunk 3 above)
Comment (frankleaf) [reply pkg/ddc/base/engine.go:46]: thx, I agree with you

File: pkg/controllers/v1alpha1/dataload/dataload_controller.go (hunk 4 above)
Comment (frankleaf) [reply pkg/controllers/v1alpha1/dataload/dataload_controller.go:132]: thx, done

**Comment by frankleaf** [comment]
> @frankleaf thanks very much for your PR. It's very important to decouple the definition and implementation. Would you mind to add some unit test here like #832 . Thank you in advanced.

ths, in action

File: pkg/controllers/v1alpha1/dataload/dataload_controller.go (hunk 5 above)
Comment (frankleaf) [reply pkg/controllers/v1alpha1/dataload/dataload_controller.go:223]: thx, forget to add remove engine when delete, already done

File: pkg/controllers/v1alpha1/dataload/dataload_controller.go [hunk 6]
```diff
@@ -91,6 +91,7 @@ func (r *DataLoadReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 
 	// 2. Reconcile deletion of the object if necessary
 	if utils.HasDeletionTimestamp(dataload.ObjectMeta) {
+		r.RemoveEngine(ctx)
 		return r.ReconcileDataLoadDeletion(ctx, targetDataload)
```
Comment (cheyang) [inline pkg/controllers/v1alpha1/dataload/dataload_controller.go:95]: I suggest to move `r.RemoveEngine` into `r.ReconcileDataLoadDeletion`. Thanks.

**Review by cheyang — CHANGES_REQUESTED** [review]

File: pkg/utils/dataset.go [hunk 7]
```diff
@@ -66,3 +69,21 @@ func GetAccessModesOfDataset(client client.Client, name, namespace string) (acce
 	return accessModes, err
 
 }
+
+// isTargetPathUnderFluidNativeMounts checks if targetPath is a subpath under some given native mount point.
```
Comment (cheyang) [inline pkg/utils/dataset.go:73]: isTargetPathUnderFluidNativeMounts should be IsTargetPathUnderFluidNativeMounts

Comment (frankleaf) [reply pkg/utils/dataset.go]: thx, done

File: pkg/controllers/v1alpha1/dataload/dataload_controller.go (hunk 6 above)
Comment (frankleaf) [reply pkg/controllers/v1alpha1/dataload/dataload_controller.go]: thx, done

**Comment by cheyang** [comment]
@frankleaf please fix the conflict, thanks.

**Review by cheyang — APPROVED** [review]

**Pull request merged**